### DISCUSSION TITLE
Posting to a CFC Not Working for Form | Wheels.RouteNotFound

### DISCUSSION THREAD

**Discussion by BrevardCounty-Wade** [wheels-dev/wheels#1211] [Q&A] [answered]
Hello,

I am having some trouble with updating a ColdFusion server to a Lucee + CFWheels setup. The issue was not encountered when running the project on Lucee without CFWheels. However, CFWheels has been rather helpful in its development, and I'd like to keep its inclusion as a framework.

Most of the web app transferred from that pure Lucee version with Tomcat URL rewriting. However, when I tried to include CFWheels in a separate new version, only one operation did not perform as expected. It involves a form with a jQuery 'post()' method using JSON for form elements to be arguments in a CFC file. It would consistently return a Wheels.ViewNotFound error. I did include Tomcat URL rewriting, and I have a temporary setup for wildcard routing with GET and POST methods.

After updating the URL path to match the planned changes and accounting for the suggested tomcat URL rewriting in documentation I had done earlier, the error updated to a Wheels.RouteNotFound error. I have no other alterations nor config settings that would interfere with the POST method. I have rewritten the jQuery operation to be a CFML condition using CGI scope for the time being, and I did a 'cfinvoke' function to the CFC file with the necessary arguments and values.  Since then, I receive a view updating from the form page to rewrite.cfm with the following error:

**Wheels.RouteNotFound**
Incorrect HTTP Verb for route

**Suggested action**
The path does not allow POST requests, only GET requests. Ensure you are using the correct HTTP Verb and that your config/routes.cfm file is configured correctly.

 I have tried the following since that error:

1. Move the CFC from /controllers to /views/[current form view folder].
2. Include a copy of CFC file in /miscellaneous.
3. Move the CFC back to /controllers and deleted the copy in /miscellaneous
3. Made a remote function in the CFC file related to the form page and receiving page.
4. Attempted to use the verifies() method in a config() function to allow for POST methods.

I have experienced success with the other CFC files using 'cfinvoke', but this form is the only one that post to a CFC to then use its plain text to make a 'cfswitch' decision. It would proceed to the next view on 'cflocation' request. 

Is there anything else I need to include besides the URL in that 'cflocation' statement, or is the error I am encountering something else? I'm lost at the moment to what is preventing me from doing this. I hope this helps, but let me know if you need additional information for the issue.

**Update:** I moved the operations from the CFC into a post method operation. The error occurs, but, at least, it wasn't an issue with the CFC and returning JSON.

**Comment by chapmandu** [comment]
I sounds like you're missing a post route configuration in your /config/routes.cfc, It might look something like:

`.post(name="someRouteName" to="Users##create", pattern="/users/create")`

The pattern should match the URL you're posting to.

There is lots of info about routing here.. https://guides.cfwheels.org/cfwheels-guides/handling-requests-with-controllers/routing

I hope this helps..

**Reply by BrevardCounty-Wade** [reply]
I have the post route configured now to the 'controllers' folder. I see it in View Routes page, and the name of the route matches the event. Although, the error still occurs.

Edit: I also tried resources() function for the controllers folder and receiving page for the cflocation function for testing purposes. Error persists.

**Comment by dbelanger** [comment, marked as answer]
Hey @BrevardCounty-Wade , normally your initial error "Wheels.ViewNotFound" has nothing to do with routing at all.  It usually means that Wheels processed the controller and then didn't know what to do after.  If you don't specify that a view isn't required, by default, it looks for a file of the same name in the view folder.  When it can't find it, Wheels simply says "view not found".

A little more detail on your points above:

1. You should **never** have a cfc in a views folder.  The views folder are always cfm files because we're returning web pages, excels or pdfs.  If you're returning straight JSON, you don't even need a view!  You can tell the controller to simply return json and pass in the data you'd like converted to json (here's the [link](https://guides.cfwheels.org/cfwheels-guides/handling-requests-with-controllers/responding-with-multiple-formats#responding-to-different-formats-in-the-controller) in the docs for that).
2. The miscellaneous folder is a good place to put files you don't want to go through the Wheels router. Personally, I don't use it very much at all.
4. Remote functions are no longer considered a good practice in CF.  We used to be heavy on the WSDL but APIs are easily to build, secure and return the info you need
5. Verifies is like a second layer of security but since version 2, the router pretty much ensures which requests are allowed and which aren't.

**Reply by BrevardCounty-Wade** [reply]
Thank you for these details. I planned to perform it as originally intended from the ColdFusion version, but I believe I may need to update the authentication method. Yes, straight JSON is returned. I may need to update the process to be inside the view's controller (which is where the attempt with the remote function came into testing). For now, I am receiving Wheels.RouteNotFound as I am left on rewrite.cfm after submitting the form.
